### PR TITLE
On DDL operation keep the lock on partitions until end of transaction

### PR DESCRIPTION
**Pull request by maksm90**
Commit fa287d0 have introduced early lock releasing on partitions within operations with partitioned parent table. This was done to save memory in lock manager I suppose. But such behavior incurs some issues, e.g., possible segfault when we perform drop of partitioned table and concurrent `pg_total_relation_size()` call on some AO partition.

Obviously, we have to restore full locking of inheritance hierarchy as it's done in vanilla PostgreSQL. To eliminate running out of lock memory we need to increase max_locks_per_transactions setting. Memory is cheap, and if we're working with tens of thousands of partitions, we can afford reserving some memory for the lock manager.

Partially for DML operations the problem of partitions locking was resolved in commit 99fbc8e. Therefore in current work the support for DDL operations have been added. And it's in fact backport of corresponding commit to master 86ded36. Also there have been added test scenario to reproduce noted above segfault case and related assert check inside `calculate_table_size()` function.

Backport of PR https://github.com/greenplum-db/gpdb/pull/10468 .
Fixes https://github.com/greenplum-db/gpdb/issues/12962 . All details about faulty case you might find there.

## Here are some reminders before you submit the pull request
- [ ] Add tests for the change
- [ ] Document changes
- [ ] Communicate in the mailing list if needed
- [ ] Pass `make installcheck`
- [ ] Review a PR in return to support the community

### CODE REVIEW

**Review by seryozhasmirnov — DISMISSED** [review]

File: src/test/isolation2/sql/ddl_on_root_locks_all_parts.sql [hunk 1]
```diff
@@ -0,0 +1,62 @@
+-- For DDL op running over partitioned tables table we should acquire
```
Comment (InnerLife0) [inline src/test/isolation2/sql/ddl_on_root_locks_all_parts.sql]: Doubled `table(s)`.

File: src/test/isolation2/sql/ddl_on_root_locks_all_parts.sql [hunk 2]
```diff
@@ -0,0 +1,62 @@
...
+-- AccessExclusive locks on root relation and child partitions and keep their
```
Comment (InnerLife0) [inline src/test/isolation2/sql/ddl_on_root_locks_all_parts.sql]: Probably missed `locks` word after `their` or `their` and `them` mixed up.

File: src/test/isolation2/sql/ddl_on_root_locks_all_parts.sql [hunk 3]
```diff
@@ -0,0 +1,62 @@
...
+-- pg_aoseg relation for some partition on some segment node if its progress is
+-- not restricted by locks on partitions from concurrent DROP stmt.
+-- After fixing of partition locking issue the second session have to stuck on
+-- acquiring of AccessShare lock and take advance only after commit in the first
+-- session completes.
+SELECT gp_inject_fault('pg_calculate_table_size_before_pg_aoseg_estimation', 'suspend', dbid)
+FROM gp_segment_configuration WHERE role = 'p' AND content = 0;
+2&: SELECT pg_total_relation_size('ptest_1_prt_1'::regclass);
+
+-- Wait some amount time so that the second session have achieved fault point if
```
Comment (InnerLife0) [inline src/test/isolation2/sql/ddl_on_root_locks_all_parts.sql]: `_of_ time`

File: src/test/isolation2/sql/ddl_on_root_locks_all_parts.sql [hunk 4]
```diff
@@ -0,0 +1,62 @@
...
+-- it has progress
+SELECT pg_sleep(1);
+
+-- Try to stop txn commit on some segment in the first session up to catalog
+-- cache invalidation but after the routine that removes PGPROC entry from
+-- ProcArray so that renewed snapshots instantiated in concurrent transactions
+-- have treated current transaction as completed
```
Comment (InnerLife0) [inline src/test/isolation2/sql/ddl_on_root_locks_all_parts.sql]: Sorry for much remarks to comments, especially from the man who's not the best in English too, but... 
It's really hard to read such long 4-row sentence without any comma. We can do `, but` or `, so` or break one long sentence on two short (which we can do most often).

Comment (maksm90) [reply src/test/isolation2/sql/ddl_on_root_locks_all_parts.sql]: Thanks, comment message really looks ugly. Tried to refine it as far as I can.

**Review by InnerLife0 — APPROVED** [review]

**Pull request merged**